src/System.Diagnostics.Tracing.Telemetry/src/System/Diagnostics/Tracing/TelemetryListener.cs

```diff
@@ -0,0 +1,249 @@
+            var lockObj = DefaultListener;
+            if (lockObj == null)
+                lockObj = this;
+
+            // Insert myself into the list of all Listeners.   
+            lock (lockObj)
+            {
+                // Issue the callback for this new telemetry listener. 
+                var callback = s_allListenersCallback;
+                if (callback != null)
+                    callback(this);
+
+                // And add it to the list of all past listeners.  
+                m_next = s_allListeners;
+                s_allListeners = this;
+            }
+        }
+
+        /// <summary>
+        /// Clean up the NotificationListeners.   Notification listeners do NOT DIE ON THEIR OWN
+        /// because they are in a global list (for discoverability).  You must dispose them explicitly. 
+        /// Note that we do not do the Dispose(bool) pattern because we frankly don't want to support
+        /// subclasses that have non-managed state.   
+        /// </summary>
+        virtual public void Dispose()
+        {
+            // Remove myself from the list if all listeners.  
+            lock (DefaultListener)
+            {
+                m_disposed = true;
+                if (s_allListeners == this)
+                    s_allListeners = s_allListeners.m_next;
+                else
+                {
+                    var cur = s_allListeners;
+                    while (cur != null)
+                    {
+                        if (cur.m_next == this)
+                        {
+                            cur.m_next = this.m_next;
+                            break;
+                        }
+                        cur = cur.m_next;
+                    }
+                }
+                m_next = null;
+            }
+
+            // TODO audit for races...  Should this be under the lock above?  
+
+            // Indicate completion to all subscribers.  
+            Subscription subscriber = null;
+            Interlocked.Exchange(ref subscriber, m_subscriptions);
+            while (subscriber != null)
+            {
+                subscriber.Observer.OnCompleted();
+                subscriber = subscriber.Next;
+            }
+            // The code above also nulled out all subscriptions. 
+            // TODO: should we block new subscriptions after we have been disposed?  
+        }
+
+        /// <summary>
+        /// Return the name for the ToString() to aid in debugging.  
+        /// </summary>
+        /// <returns></returns>
+        public override string ToString()
+        {
+            return Name;
+        }
+
+        #region private
+
+        // NotificationSource implementation
+        public override bool IsEnabled(string telemetryName)
+        {
+            for (var curSubscription = m_subscriptions; curSubscription != null; curSubscription = curSubscription.Next)
```

Comment:
Other usage of m_subscriptions involves interlockeds... do you want to do a volatile read?  Or not necessary?